Gas refund changes with NEP-536 (#2549)
near/docs: docs/protocol/gas.md

The only exception to this rule is when you make a function call to a contract. In this case, you need to define how many gas units to use, up to a maximum value of `300Tgas`. This amount will be converted to $NEAR using the network's gas price and deducted from your account's balance.

If the transaction ends up using less gas than the amount deducted, the difference will simply be **refunded to your account**. However, the network subtracts a **gas refund fee** of at leat 1 Tgas or up to 5% of the unspent gas.

<details>

<summary> Gas Refund Fee </summary>

Since protocol version 78, the unspent gas at the end of receipt execution is subject to a gas refund fee. The fee is calculated as `max(1 Tgas, 0.05 * unspent_gas) * gas_price`. The gas price is from the time of purchase.

> But why introduce such a fee, when refunding 100% of unspent gas allows developers to just attach the maximum amount of gas and not worry about gas estimations too much?

Reviewer: ```suggestion
_But why introducing such a fee instead of refunding all gas?_
```

> The resason is that attaching too much gas to function calls makes the network less efficient.

Reviewer: ```suggestion
The reason is that attaching too much gas to function calls makes the network less efficient.
```

> Especially congestion control between shards is tricky when transactions have much more gas attached than they actually use. The network limits how many cross contract calls can target a single shard per chunk, to avoid huge queues of incoming receipts on a shard. The limit looks at attached gas as an upper boundary for how much work the receipt causes on the receiving shard. Attaching too much gas can cause this limit to be too restrictive, which stalls shards unnecessarily.

Reviewer: ```suggestion
Congestion control between shards gets tricky when transactions have much more gas attached than they actually use. The network limits how many cross contract calls can target a single shard per chunk, to avoid huge queues of incoming receipts on a shard. This limit sees the attached gas as an upper boundary for how much work the receipt causes on the receiving shard. Attaching too much gas can cause this limit to become too restrictive, which stalls shards unnecessarily.
```

_But why introducing such a fee instead of refunding all gas?_

The reason is that attaching too much gas to function calls makes the network less efficient.

Congestion control between shards gets tricky when transactions have much more gas attached than they actually use. The network limits how many cross contract calls can target a single shard per chunk, to avoid huge queues of incoming receipts on a shard. This limit sees the attached gas as an upper boundary for how much work the receipt causes on the receiving shard. Attaching too much gas can cause this limit to become too restrictive, which stalls shards unnecessarily.

The other inefficiency comes from the refund receipts that prior to version 78 have been created for essentially every function call. While each of them is relatively cheap to execute, in the sum they were a significant part of the global traffic on NEAR Protocol.

To read more about the deicsion to introduce this fee, pleaes take a look at the [NEP-536](https://github.com/near/NEPs/pull/536)

</details>

<details>

:::tip

> In other chains, paying a higher gas price gets your transaction processed faster. In NEAR, **gas costs are deterministic**, and you **can't pay to get priority**. Any extra gas attached to a transaction is refunded, minus some fee for attaching unnecesary gas.

Reviewer: ```suggestion
In other chains, paying a higher gas price gets your transaction processed faster. In NEAR, **gas costs are deterministic**, and you **can't pay to get priority**. Any extra gas attached to a transaction is refunded, minus some fee for attaching unnecessary gas.
```
